### DISCUSSION TITLE
additional repo "sync" request types

### DISCUSSION THREAD

**Discussion by bnewbold** [bluesky-social/atproto#1158] [Protocol (atproto)]
Separate from whatever federation scheme gets used for streaming replication between instances, here are a couple low-level repo sync requests that seem like they would be useful. In all of these cases, the client could just fetch the full repo history and compute from that, which would require less "trust" in the PDS, but be inefficient.

- given a repo (DID) and list of paths (collection/TID pairs), return the records plus the commit/root blocks and any MST tree nodes needed to verify the signature down to those records. useful for clients to verify records without fetching full repo state
- fetch the full tree (commit/root, MST nodes, and all records) for repo, but only at the current commit (don't iterate back through the full IPLD DAG including previous commits). this is all most non-archival indexers and clients would want. maybe also filter to specific collections?
- fetch full history of an AT URI for given repo: returns an ordered list of `{ commit, record }` pairs, indicating the create/update/delete history of that key. only for history accessible from the current commit "previous" chain (aka, removing history by re-writing this chain is possible). use case is for clients to understand the log of edits to a record
- variant of `com.atproto.repo.listRecords` which allows giving just a TID range (open or closed), but not specify collection. could be used to return recent records of any type

**Comment by dholms** [comment]
Hey good timing, I was actually getting ready to dive back into our sync primitives today & both of your issues are relevant to it. If you'd believe it, `com.atproto.sync.getRepo` route hasn't gotten any love since the ADX days (May 😅).

Already have plans for 1 & 2. 

I like the sound of 3 but may hold off for now since we don't support edits in the general case yet. Curious if you have a particular use case in mind?

For 4, I'm worried that this could be misleading for a coupld reason: 
- TIDs are not _required_ to be added in order. in some sense they are "just strings", but they're strings designed to give values rough time-ordering with in a collection. 
- records don't need to be keyed by TID. some collection (profile & declaration for instance) are keyed by a literal (`self`) since there is only one value there. We haven't found a usecase yet, but it's not out of the question that some collections will have other types of keys

Allowing queries over a TID range sort of elevates TIDs into something that they aren't quite. The failure case here is that a client assumes they're getting all records in a repo by using TIDs as a cursor, but they're not getting out of order TIDs or non-tid-keyed records

**Comment by bnewbold** [comment]
For 3, the use case is a UX for understanding edits. Eg, today I got a notification on mastodon that a user had edited a post that I had boosted (reposted) recently. But there was no way to see what actually changed about the post. Would be nice to have a view which showed diffs, or are least all the previous versions.

Not urgent at the API layer because clients can just pull the full history if they really want to implement this kind of view, it is just an efficiency thing. And that goes for all of these, really.

For 4, :+1:

**Comment by DavidBuchanan314** [comment]
I assume the second one is what became `getCheckout` ?

**Reply by bnewbold** [reply]
yup!
